common/src/Model/Tag.h

```diff
@@ -0,0 +1,305 @@
+         * whereby the type is just a unique unsigned integer that acts as a bitmask index.
+         *
+         * Furthermore, a tag can have attributes.
+         */
+        class Tag {
+        public:
+            using TagType = unsigned long;
+        protected:
+            TagType m_type;
+            String m_name;
+            std::vector<TagAttribute> m_attributes;
+
+            /**
+             * Creates a new tag with the given type, name and attributes. The tag's type will be set automatically.
+             *
+             * @param type the tag's type
+             * @param name the tag's name
+             * @param attributes the tag's attributes
+             */
+            Tag(TagType type, String name, std::vector<TagAttribute> attributes);
+        public:
+            /**
+             * Creates a new tag with the given name and attributes. The tag's type will be set automatically.
+             *
+             * @param name the tag's name
+             * @param attributes the tag's attributes
+             */
+            Tag(String name, std::vector<TagAttribute> attributes);
+
+            /**
+             * Returns the type of this tag.
+             */
+            TagType type() const;
+
+            /**
+             * Sets the tag type. Can be called only once.
+             *
+             * @param type the type
+             */
+            void setType(TagType type);
+
+            /**
+             * Returns the name of this tag.
+             */
+            const String& name() const;
+
+            /**
+             * Returns the attributes of this tag.
+             */
+            const std::vector<TagAttribute>& attributes() const;
+
+            friend bool operator==(const Tag& lhs, const Tag& rhs);
+            friend bool operator<(const Tag& lhs, const Tag& rhs);
+        };
+
+        /**
+         * Tag references are used to store the tags attached to an object implementing the Taggable interface.
+         */
+        class TagReference {
+        private:
+            const Tag& m_tag;
+        public:
+            /**
+             * Creates a new reference to the given tag.
+             *
+             * @param tag the referenced tag
+             */
+            explicit TagReference(const Tag& tag);
+
+            /**
+             * Returns the referenced tag.
+             */
+            const Tag& tag() const;
+
+            friend bool operator==(const TagReference& lhs, const TagReference& rhs);
+            friend bool operator<(const TagReference& lhs, const TagReference& rhs);
+        };
+
+        class TagManager;
+        class TagMatcher;
+
+        /**
+         * Implementing this interface gives a class the ability to be tagged.
+         */
+        class Taggable {
+        private:
+            Tag::TagType m_tagMask;
+            std::set<TagReference> m_tags;
```

Comment:
I don't like TagReference because it is code that does nothing, and hides the dangers of this setup - this should just be `std::set<Tag*> m_tags;` which communicates the design more clearly. To me this is much more important than preventing nulls from being in the std::set which I assume is the motivation for having TagReference?

More importantly, anything implementing Taggable now has a hidden dependency on the lifetime of the Tag objects outliving all Taggables using those tags. This is fine, but we should mention that in the class docs for Taggable. The class docs should also explain where the Tags are owned. It looks like SmartTags are owned by TagManager; what about other tags?

Reply:
I somehow missed these comments. All tags are owned by the tag manager. `TagReference` was introduced for two reasons: I wanted a reference instead of a pointer because it can never be null, and I want to be able to add code in the future that will be executed whenever a tag reference is created / destroyed, e.g. a usage counter mechanism, or updating a cache of the tagged objects in the tag itself. This will be important in the future for more advanced usage of smart tags such as smart layers and such.